Edit Round function in MathTrig.cs

pdennis33 opened the pull request:
The widely accepted rounding in math is to round .5 to 1 instead of 0.  Excel does this by default, but the current function Math.Round(value, digits) rounds .5 to 0.  I've added the 3rd parameter MidpointRounding.AwayFromZero to make it act like Excel does by default, and round .5 up to 1.


#### What's this PR do?
#### Where should the reviewer start?
#### How should this be manually tested?
#### Any background context you want to provide?

#### Screenshots (if appropriate)
#### Questions:
- Is there a blog post?
- Does the knowledge base need an update?
- Does this add new (C#) dependencies?

- [ ] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

igitur commented:
Looks good. But please add a unit test. Thanks.

pdennis33 commented:
Sorry, I'm a bit ashamed to admit this, but while I've coded extensively, I have not used unit testing nor GitHub.  I'd be happy to add a unit test.  Is it ok to please point me in the direction of where I can read how to add a unit test on here?  I can check how to actually do the unit test with Visual Studio but I'm not sure how you'd prefer it be added here.  Thank you!  Sorry for not being up on that stuff :)

pdennis33 commented:
Ok I think and hope this is what you were looking for.  I added a unit test, and I'm pasting my results below from the old method which fails first and the new method which passes.  I'll also show a screenshot of how I found this issue and the numbers used.  Everything was identical except changing the Round function as I proposed.

Test Results from **old** Round function:
```
Test Name:	MathTrig_RoundShouldRoundAwayFromZero
Test FullName:	ClosedXML_Tests.Excel.CalcEngine.MathTrigTests.MathTrig_RoundShouldRoundAwayFromZero
Test Source:	C:\Users\Paul\Downloads\ClosedXML-master\ClosedXML_Tests\Excel\CalcEngine\MathTrigTests.cs : line 12
Test Outcome:	Failed
Test Duration:	0:00:00.8746322

Result StackTrace:	at ClosedXML_Tests.Excel.CalcEngine.MathTrigTests.MathTrig_RoundShouldRoundAwayFromZero() in C:\Users\Paul\Downloads\ClosedXML-master\ClosedXML_Tests\Excel\CalcEngine\MathTrigTests.cs:line 22
Result Message:	Assert.AreEqual failed. Expected:<31.57>. Actual:<31.56>.
```

Test Results from **proposed** Round function:
```
Test Name:	MathTrig_RoundShouldRoundAwayFromZero
Test FullName:	ClosedXML_Tests.Excel.CalcEngine.MathTrigTests.MathTrig_RoundShouldRoundAwayFromZero
Test Source:	C:\Users\Paul\Downloads\ClosedXML-master\ClosedXML_Tests\Excel\CalcEngine\MathTrigTests.cs : line 12
Test Outcome:	Passed
Test Duration:	0:00:01.0090358
```

![image](https://user-images.githubusercontent.com/29433143/27187656-608c01f8-51b1-11e7-89ca-bfaed25c33cf.png)

pdennis33 commented:
@igitur I deleted my comment about the guidance on unit testing, since I hope this is a good example of what you were looking for.  Thanks!

igitur commented:
Thanks! Sorry, you beat me before I could explain the unit tests. 

Firstly, I notice you based the PR on the master branch. You should base it on the develop branch. I think you should be able to rebase without too much effort. Let me know if you need help and sorry for the upfront confusion. I should probably spec some developer guidelines. The benefit of this is that our CI will automatically build and run the tests and you will be able to see the result here on this issue page.

The unit tests is on the right track, but I can immediately see it's giong to fail, because referencing a local file on your `D:` drive. For unit tests for formulas, you can check out the other unit tests. You don't even need to create or open an existing workbook. You can use `XLWorkbook.Evaluate` directly on a formula, e.g. :+1: 
```c#

        [Test]
        public void Count()
        {
            var ws = workbook.Worksheets.First();
            int value;
            value = ws.Evaluate(@"=COUNT(D3:D45)").CastTo<int>();
            Assert.AreEqual(0, value);

            value = ws.Evaluate(@"=COUNT(G3:G45)").CastTo<int>();
            Assert.AreEqual(43, value);

            value = ws.Evaluate(@"=COUNT(G:G)").CastTo<int>();
            Assert.AreEqual(43, value);

            value = workbook.Evaluate(@"=COUNT(Data!G:G)").CastTo<int>();
            Assert.AreEqual(43, value);
        }
```

Also note we're using `NUnit3` for unit tests, not `MSTest`. You can install the `NUnit 3 Test Adapter` if you want to run your tests inside Visual Studio. Info for installation at https://github.com/nunit/docs/wiki/Adapter-Installation

igitur commented on ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
```
Use NUnit

igitur commented on ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs:
```diff
@@ -0,0 +1,26 @@
+            double cellAmount = 31.565d;
+            double roundedAmount = Math.Round(cellAmount, 2, MidpointRounding.AwayFromZero);
+
+            using (XLWorkbook testBook = new XLWorkbook(@"D:\Test.xlsx"))
```
No need to open a workbook. `XLWorkbook.EvaluateExpr`

igitur commented on ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs:
```diff
@@ -0,0 +1,26 @@
+            }
+        }
+    }
+}
```
Try to add as many use cases as you can think. Include negative numbers too, for example.

vbjay commented:
No need to install extension.
https://www.nuget.org/packages/NUnit3TestAdapter

On Thu, Jun 15, 2017, 11:37 AM Francois Botha <notifications@github.com>
wrote:

> *@igitur* requested changes on this pull request.
> ------------------------------
>
> In ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs
> <https://github.com/ClosedXML/ClosedXML/pull/349#discussion_r122235358>:
>
> > @@ -0,0 +1,26 @@
> +using System;
> +using Microsoft.VisualStudio.TestTools.UnitTesting;
>
> Use NUnit
> ------------------------------
>
> In ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs
> <https://github.com/ClosedXML/ClosedXML/pull/349#discussion_r122235447>:
>
> > +using System;
> +using Microsoft.VisualStudio.TestTools.UnitTesting;
> +using ClosedXML.Excel;
> +
> +namespace ClosedXML_Tests.Excel.CalcEngine
> +{
> +    [TestClass]
> +    public class MathTrigTests
> +    {
> +        [TestMethod]
> +        public void MathTrig_RoundShouldRoundAwayFromZero()
> +        {
> +            double cellAmount = 31.565d;
> +            double roundedAmount = Math.Round(cellAmount, 2, MidpointRounding.AwayFromZero);
> +
> +            using (XLWorkbook testBook = new XLWorkbook(@"D:\Test.xlsx"))
>
> No need to open a workbook. XLWorkbook.EvaluateExpr
> ------------------------------
>
> In ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs
> <https://github.com/ClosedXML/ClosedXML/pull/349#discussion_r122235729>:
>
> > +        public void MathTrig_RoundShouldRoundAwayFromZero()
> +        {
> +            double cellAmount = 31.565d;
> +            double roundedAmount = Math.Round(cellAmount, 2, MidpointRounding.AwayFromZero);
> +
> +            using (XLWorkbook testBook = new XLWorkbook(@"D:\Test.xlsx"))
> +            {
> +                IXLWorksheet testSheet;
> +                testBook.TryGetWorksheet("Sheet1", out testSheet);
> +
> +                double actualCellAmount = testSheet.Cell(2, 2).GetDouble();
> +                Assert.AreEqual(roundedAmount, actualCellAmount);
> +            }
> +        }
> +    }
> +}
>
> Try to add as many use cases as you can think. Include negative numbers
> too, for example.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/349#pullrequestreview-44334689>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsb-lFYR9ZWNy9Dh3w72TgTAflgKnks5sEU-igaJpZM4N6AyD>
> .
>

pdennis33 commented:
Thanks a lot guys!  @igitur  Thanks for the help!  I hope I made the changes correctly.  I changed the pull request to be based on the develop branch.  I see it says it has conflicts that must be resolved but maybe that's normal?  I realized there was already a Round() tests within FunctionTests.cs so I just added the example from my spreadsheet that was causing the test to fail.  It had 7 tests already, so I figured instead of making a brand new one, I'd just add an 8th test.  Hopefully this satisfies those changes.  Thanks again!

pdennis33 commented:
Maybe I need to delete this pull request and create a new fork based on the develop branch and make the changes on there and then create a new pull request on that?

vbjay commented:
You just need to rebase and push your changes to the same branch.

On Thu, Jun 15, 2017 at 2:26 PM pdennis33 <notifications@github.com> wrote:

> Maybe I need to delete this pull request and create a new fork based on
> the develop branch and make the changes on there and then create a new pull
> request on that?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/349#issuecomment-308817473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsSPC_Ok3CQyABv_QOT7wbv_UdyzVks5sEW3dgaJpZM4N6AyD>
> .
>

vbjay commented:
You will have to push with force-with-lease

On Thu, Jun 15, 2017 at 2:47 PM Jay Asbury <vbjay.net@gmail.com> wrote:

> You just need to rebase and push your changes to the same branch.
>
> On Thu, Jun 15, 2017 at 2:26 PM pdennis33 <notifications@github.com>
> wrote:
>
>> Maybe I need to delete this pull request and create a new fork based on
>> the develop branch and make the changes on there and then create a new pull
>> request on that?
>>
>> —
>> You are receiving this because you commented.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ClosedXML/ClosedXML/pull/349#issuecomment-308817473>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADdhsSPC_Ok3CQyABv_QOT7wbv_UdyzVks5sEW3dgaJpZM4N6AyD>
>> .
>>
>

pdennis33 commented:
Thanks @vbjay but I'm sorry, I don't quite follow.  I'm new to Git and GitHub so I apologize.  I clicked on Edit up top and changed the base of the Pull Request to the develop branch.  However, I'm not sure what you mean by pushing my changes to the same branch and also push with force-with-lease.  Thank you!

igitur commented:
Superceded by #353
